codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parse/JsonParserGenerator.kt

```diff
@@ -0,0 +1,457 @@
+import software.amazon.smithy.rust.codegen.util.getTrait
+import software.amazon.smithy.rust.codegen.util.hasTrait
+import software.amazon.smithy.rust.codegen.util.outputShape
+import software.amazon.smithy.rust.codegen.util.toPascalCase
+import software.amazon.smithy.utils.StringUtils
+
+class JsonParserGenerator(
+    protocolConfig: ProtocolConfig,
+    private val httpBindingResolver: HttpBindingResolver,
+) : StructuredDataParserGenerator {
+    private val model = protocolConfig.model
+    private val symbolProvider = protocolConfig.symbolProvider
+    private val runtimeConfig = protocolConfig.runtimeConfig
+    private val smithyJson = CargoDependency.smithyJson(runtimeConfig).asType()
+    private val codegenScope = arrayOf(
+        "Error" to smithyJson.member("deserialize::Error"),
+        "expect_blob_or_null" to smithyJson.member("deserialize::token::expect_blob_or_null"),
+        "expect_bool_or_null" to smithyJson.member("deserialize::token::expect_bool_or_null"),
+        "expect_document" to smithyJson.member("deserialize::token::expect_document"),
+        "expect_number_or_null" to smithyJson.member("deserialize::token::expect_number_or_null"),
+        "expect_start_array" to smithyJson.member("deserialize::token::expect_start_array"),
+        "expect_start_object" to smithyJson.member("deserialize::token::expect_start_object"),
+        "expect_string_or_null" to smithyJson.member("deserialize::token::expect_string_or_null"),
+        "expect_timestamp_or_null" to smithyJson.member("deserialize::token::expect_timestamp_or_null"),
+        "expect_unescaped_string_or_null" to smithyJson.member("deserialize::token::expect_unescaped_string_or_null"),
+        "json_token_iter" to smithyJson.member("deserialize::json_token_iter"),
+        "Peekable" to RuntimeType.std.member("iter::Peekable"),
+        "skip_value" to smithyJson.member("deserialize::token::skip_value"),
+        "Token" to smithyJson.member("deserialize::Token"),
+    )
+
+    override fun payloadParser(member: MemberShape): RuntimeType {
+        val shape = model.expectShape(member.target)
+        check(shape is UnionShape || shape is StructureShape) { "payload parser should only be used on structures & unions" }
+        val fnName = symbolProvider.deserializeFunctionName(shape) + "_payload"
+        return RuntimeType.forInlineFun(fnName, "json_deser") {
+            it.rustBlockTemplate(
+                "pub fn $fnName(input: &[u8]) -> Result<#{Shape}, #{Error}>",
+                *codegenScope,
+                "Shape" to symbolProvider.toSymbol(shape)
+            ) {
+                rustTemplate(
+                    """
+                    let mut tokens_owned = #{json_token_iter}(input).peekable();
+                    let tokens = &mut tokens_owned;
+                    """,
+                    *codegenScope
+                )
+                rust("let result =")
+                deserializeMember(member, false)
```

Comment:
I'd probably create a `Ctx` for this, partially just to follow the pattern of all the other deserializers & partially to make the code a little easier to read without jumping to definition to see what `false` means

Reply:
I may end up having to introduce context to correctly handle the non-sparse list nullability issue. Otherwise, it seemed like overkill for one context value. I should have used named arguments though to clarify the boolean value.